src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -73,6 +74,28 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
     },
   };
 
+  function checkShouldComponentUpdate(workInProgress, oldProps, newProps, newState) {
+    if (oldProps === null ||
+        (workInProgress.updateQueue && workInProgress.updateQueue.isForced)) {
+      return true;
+    }
+
+    const instance = workInProgress.stateNode;
+    if (typeof instance.shouldComponentUpdate === 'function') {
+      return instance.shouldComponentUpdate(newProps, newState);
+    }
+
+    const type = workInProgress.type;
+    if (type.prototype && type.prototype.isPureReactComponent) {
```

Comment:
However, this can explode number of code paths so I figured we might want to optimize for collections of common patterns. For example: isPureComponent + no life-cycles. isPureComponent + only componentDidUpdate/componentWillUnmount but no other life-cycles. That way we can branch early and not do any of the checks. Never schedule update effects etc.